test/functional/ui/incsubstitution_spec.lua

```diff
@@ -0,0 +1,805 @@
+      ~                   {10:|}~                  |
+      ~                   {10:|}~                  |
+      ~                   {10:|}~                  |
+      ~                   {10:|}~                  |
+      {11:[No Name] [+]        }{10:[No Name] [+]      }|
+      Inc substitution on                     |
+      two lines                               |
+                                              |
+      ~                                       |
+      ~                                       |
+      {10:[No Name] [+]                           }|
+       [2]two lines                           |
+                                              |
+      ~                                       |
+      ~                                       |
+      ~                                       |
+      ~                                       |
+      ~                                       |
+      {10:[inc_sub]                               }|
+      :%s/tw^                                  |
+    ]])
+  end)
+
+  local settings = {
+  "splitbelow",
+  "splitright",
+  "noequalalways",
+  "equalalways eadirection=ver",
+  "equalalways eadirection=hor",
+  "equalalways eadirection=both",
+  }
+
+  for _, setting in pairs(settings) do
+    it("are not affected by " .. setting, function()
+
+      execute("set " .. setting)
+
+      feed(":%s/tw")
+      screen:expect([[
+        Inc substitution on                     |
+        two lines                               |
+                                                |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        {11:[No Name] [+]                           }|
+         [2]two lines                           |
+                                                |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        ~                                       |
+        {10:[inc_sub]                               }|
+        :%s/tw^                                  |
+      ]])
+    end)
+  end
+
```

Comment:
Forgot a thing: since you are using undo there should be tests with &undolevels=0, &undolevels=1 and &undolevels=2 (tests should also check that existing behaviour is preserved; specifically with =0 it should undo/redo substitution, with =1 it should undo only substitution and state that it is at last change, with =2 it should be able to undo substitution and with =2 it should be able to undo text substitution and text insert then state that it is at last change; cancelling incsubstitute should also work in all cases like if incsubstitute was not enabled regarding undo behaviour) and &undolevels=-1 (tests should also check that undo does not work after incsub finishes).